### DISCUSSION TITLE
User permission & document access by uid

### DISCUSSION THREAD

**Discussion by anandsubbu007** [serverpod/serverpod#165] [Q&A]
I'm currently using appwrite for my project. The security features and document/collection level access based on user & group of user helps a lot to maintain data privacy.

Serverpod seems really interesting & I want to change my project backed to serverpod.

Is this feature Available in serverpod?
Document & table access based on user session?

**Comment by vlidholt** [comment]
Serverpod has access control features through the `auth` module. You can also add access control to individual endpoints, by setting the `scope`. You can grant access to different scopes for different users, e.g. to make a user an admin or grant them access to specific paid features.

**Reply by anandsubbu007** [reply]
ok Thanks.

**Reply by anandsubbu007** [reply]
how to create scope for users? I can't able to. find in documentation?

**Reply by anandsubbu007** [reply]
how do session scope and user scope work?

do I need to fetch for user details to get the scope?

**Comment by CordMemescape** [comment]
The OP's original question wasn't actually answered. While controlling access via an endpoint is useful to separate out Admin concerns from User concerns at a broader level, the original question was about managing access to data at an individual user level. So for example, having a collection/table called Posts and only allowing the owner of each post to see their own data.

Obviously, the data can be filtered by shaping your query such that only the user's posts are returned, but there is no implicit security to this approach. A malicious actor can 'emulate' your API call and change the query to return all posts, regardless of ownership.

Document-level (or row-level) permissions mean that the access is set at the data level - only User X has permission to User X's data. So even if the query is intercepted and changed en route to the server, no data will ever be returned that User X does not have permission to see. There is an inherent security to this model over simply filtering by query.

postgres natively supports row-level security so if this isn't already supported by serverpod, then I would imagine that it shouldn't be too difficult to implement. However, not being familiar with the below, I cannot say how flexible it would be to add new user or user groups to the row-level security. In my current set-up, I can create a 'team', add a number of users to that team and then set the permissions for a group of records to that team. I can manage the read/update/delete permissions on the collection and on the document-leve per user, per team and per role. Ideally, I would like to see that in Serverpod.

https://www.postgresql.org/docs/current/ddl-rowsecurity.html

**Reply by Shiba-Kar** [reply]
Ya Supabase leverage postgres  `row-level security`

**Reply by excellencemichel** [reply]
> The OP's original question wasn't actually answered. While controlling access via an endpoint is useful to separate out Admin concerns from User concerns at a broader level, the original question was about managing access to data at an individual user level. So for example, having a collection/table called Posts and only allowing the owner of each post to see their own data.
> 
> Obviously, the data can be filtered by shaping your query such that only the user's posts are returned, but there is no implicit security to this approach. A malicious actor can 'emulate' your API call and change the query to return all posts, regardless of ownership.
> 
> Document-level (or row-level) permissions mean that the access is set at the data level - only User X has permission to User X's data. So even if the query is intercepted and changed en route to the server, no data will ever be returned that User X does not have permission to see. There is an inherent security to this model over simply filtering by query.
> 
> postgres natively supports row-level security so if this isn't already supported by serverpod, then I would imagine that it shouldn't be too difficult to implement. However, not being familiar with the below, I cannot say how flexible it would be to add new user or user groups to the row-level security. In my current set-up, I can create a 'team', add a number of users to that team and then set the permissions for a group of records to that team. I can manage the read/update/delete permissions on the collection and on the document-leve per user, per team and per role. Ideally, I would like to see that in Serverpod.
> 
> https://www.postgresql.org/docs/current/ddl-rowsecurity.html

If this is implemented in **Serverpod** it will be good feature. Like Odoo access management. Thank to Serverpod's team

**Comment by Sarrius** [comment]
I'm trying to figure out how to manage scopes in my app and I can't find any starting point where to look for information.
I went through the whole documentation and seems to me there is no default way to do this.

**Reply by vlidholt** [reply]
This should definitely be added to the documentation.

In your endpoint, override the `requiredScopes` property.

```dart
class MyEndpoint {
  Set<Scope> get requiredScopes => {Scope.admin};

  ...
}
```

Then the auth module has methods for managing a users scopes.
